collector/lib/SensorClient.cpp

```diff
@@ -0,0 +1,47 @@
+#include "SensorClient.h"
+
+#include "Logging.h"
+
+namespace collector {
+bool SensorClient::Refresh() {
+  context_ = std::make_unique<grpc::ClientContext>();
+  writer_ = DuplexClient::CreateWithReadsIgnored(&sensor::CollectorService::Stub::AsyncCommunicate, channel_, context_.get());
+  if (!writer_->WaitUntilStarted(std::chrono::seconds(30))) {
+    CLOG(ERROR) << "Signal stream not ready after 30 seconds. Retrying ...";
+    CLOG(ERROR) << "Error message: " << writer_->FinishNow().error_message();
+    writer_.reset();
+    return false;
+  }
+
+  first_write_ = true;
+  stream_active_.store(true, std::memory_order_release);
+  return true;
+}
+
+SignalHandler::Result SensorClient::SendMsg(const sensor::MsgFromCollector& msg) {
+  if (!stream_active_.load(std::memory_order_acquire)) {
+    CLOG_THROTTLED(ERROR, std::chrono::seconds(10))
+        << "GRPC stream is not established";
+    return SignalHandler::ERROR;
+  }
+
+  if (first_write_ && msg.has_process_signal()) {
+    first_write_ = false;
+    return SignalHandler::NEEDS_REFRESH;
```

Comment:
It seems that `SignalHandler::NEEDS_REFRESH` is returned when `first_write_` is true and that is the case only after `Refresh` has been called. So, `SignalHandler::NEEDS_REFRESH` is returned after a refresh has been done and returning `SignalHandler::NEEDS_REFRESH` does not result in a refresh, as the name implies. It does result in `SendMsg` being called again. In which case `first_write_` will be true and this will be skipped. I don't understand what returning here instead of sending the message accomplishes.

Reply:
This is poor naming conventions on my part. `NEEDS_REFRESH` is used when the process stream has just started and it causes collector to send all processes that are stored in the threadinfo table, at startup this sends all scraped processes, while running it will ensure any processes that were not sent as part of a disconnect to be sent on reconnection.

I probably need to come up with a better term than `Refresh` for the method called after grpc is reconnected. I'll think it over and rename the method.